### PR TITLE
pkg/trace: use go.uber.com/atomic

### PR DESCRIPTION
**Pull request by djmitche**
### What does this PR do?

Uses go.uber.com/atomic to ensure alignment of atomic access in pkg/trace

### Motivation

Alignment issues with atomic access.  See #11716, https://pkg.go.dev/sync/atomic#pkg-note-BUG and https://go101.org/article/memory-layout.html

### Additional Notes

This package uses a lot of atomics!  For structs that are basically just a list of atomics, I chose to use them inline instead of via pointers.  Some of these structs are unmarshaled from JSON, and using a pointer means that the zero value is `nil` instead of `NewUint64(0)`, which leads to a lot of fiddly nonsense.  So it was just easier to use them embedded in the struct.

### Describe how to test/QA your changes

I expect any basic regression testing for trace-agent will catch issues with this change.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: docs/dev/atomics.md [hunk 1]
```diff
@@ -48,6 +48,9 @@ If the additional pointer allocation poses an undue performance burden, include
  * that it must remain in that position; and
  * why a pointer was not suitable.
 
+Note that pointers to atomic types marshal correctly to JSON as their enclosed value.
```
Comment (kayayarai) [inline docs/dev/atomics.md:51]:
```suggestion
**Note**: Pointers to atomic types marshal correctly to JSON as their enclosed value.
```

**Review by kayayarai — APPROVED** [review]
Minor suggestion, otherwise 👍  from docs

**Review by gbbr — APPROVED** [review]

File: pkg/trace/api/api_oom_test.go [hunk 2]
```diff
@@ -13,24 +13,24 @@ import (
...
 
 	"github.com/DataDog/datadog-agent/pkg/trace/config"
 	"github.com/DataDog/datadog-agent/pkg/trace/pb"
 	"github.com/DataDog/datadog-agent/pkg/trace/testutil"
+	"go.uber.org/atomic"
 )
 
 func TestOOMKill(t *testing.T) {
-	var kills uint64
+	var kills *atomic.Uint64 = atomic.NewUint64(0)
```
Comment (gbbr) [inline pkg/trace/api/api_oom_test.go:26]: Is the type needed? Wouldn't `kills := atomic.NewUint64(0)` work?

**Review by gbbr — APPROVED** [review]

File: pkg/trace/api/api_oom_test.go (hunk 2 above)
Comment (djmitche) [reply pkg/trace/api/api_oom_test.go:26]: Oh, good point -- that was a formulaic refactor.  I'll fix.

**Review by mx-psi — APPROVED** [review]

**Comment by djmitche** [comment]
(I will merge if green)

**Pull request merged**